File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
// 세그먼트 트리 
// 업데이트, 구간합 구하기

#include <iostream>

using namespace std;

int n, m, k;
long long arr[1000001];
long long seg[4000000];

long long init(int node, int st, int end) {
	if (st == end) return seg[node] = arr[st];

	int mid = (st + end) / 2;

	return seg[node] = init(node * 2, st, mid) + init(node * 2 + 1, mid + 1, end);
}

void update(int node, int st, int end, int idx, long long diff) {
	if (st <= idx && idx <= end) seg[node] += diff;	// 구간 내
	else return;	// 구간 밖

	if (st == end) return;

	int mid = (st + end) / 2;
	update(node * 2, st, mid, idx, diff);
	update(node * 2 + 1, mid + 1, end, idx, diff);
}

long long sum(int l, int r, int node, int st, int end) {
	// 구간 내 완전히 포함
	if (l <= st && end <= r) return seg[node];

	// 구간이 하나도 안 겹침
	if (end < l || r < st) return 0;

	// 구간이 일부 겹침
	int mid = (st + end) / 2;
	return sum(l, r, node * 2, st, mid) + sum(l, r, node * 2 + 1, mid + 1, end);
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);

	cin >> n >> m >> k;

	for (int i = 1; i <= n; i++) {
		cin >> arr[i];
	}

	init(1, 1, n);

	for (int i = 0; i < m + k; i++) {
		long long a, b, c;
		cin >> a >> b >> c;

		if (a == 1) { // 업데이트
			update(1, 1, n, b, c - arr[b]);
			arr[b] = c;
		} else {	// 구간합
			cout << sum(b, c, 1, 1, n) << "\n";
		}
	}

	return 0;
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
# segtree

import sys, math
input = sys.stdin.readline
sys.setrecursionlimit(100000)

# 초기 구간합 구하기
def init(arr, tree, node, st, ed):
    if st == ed:
        tree[node] = arr[st]
        return

    init(arr, tree, node * 2, st, (st + ed) // 2)
    init(arr, tree, node * 2 + 1, (st + ed) // 2 + 1, ed)
    tree[node] = tree[node * 2] + tree[node * 2 + 1]

# 재귀로 올라오면서 값, 구간값 업데이트
def update(arr, tree, node, st, ed, idx, val):
    if idx < st or idx > ed:
        return
    if st == ed:
        arr[idx] = val
        tree[node] = val
        return

    update(arr, tree, node * 2, st, (st + ed) // 2, idx, val)
    update(arr, tree, node * 2 + 1, (st + ed) // 2 + 1, ed, idx, val)
    tree[node] = tree[node * 2] + tree[node * 2 + 1]

def query(tree, node, st, ed, left, right):
    if left > ed or right < st:
        return 0
    if left <= st and ed <= right:
        return tree[node]
    lsum = query(tree, node * 2, st, (st + ed) // 2, left, right)
    rsum = query(tree, node * 2 + 1, (st + ed) // 2 + 1, ed, left, right)
    return lsum + rsum

n, m, k = map(int, input().split())
arr = [int(input()) for _ in range(n)]

h = math.ceil(math.log2(n))
tree_size = 1 << (h+1)
tree = [0] * tree_size

init(arr, tree, 1, 0, n-1)
for i in range(m + k):
    a, b, c = map(int, input().split())
    if a ==1:
        update(arr, tree, 1, 0, n-1, b-1, c)
    else:
        print(query(tree, 1, 0, n-1, b-1, c-1))
